src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIChatClient.cs

```diff
@@ -465,145 +414,71 @@ private static ChatResponse FromOpenAIChatCompletion(ChatCompletion openAIComple
-                }
+                result.StopSequences.Add(stopSequence);
             }
+        }
 
-            if (options.AdditionalProperties is { Count: > 0 } additionalProperties)
+        if (options.Tools is { Count: > 0 } tools)
+        {
+            foreach (AITool tool in tools)
             {
-                if (additionalProperties.TryGetValue(nameof(result.AudioOptions), out ChatAudioOptions? audioOptions))
-                {
-                    result.AudioOptions = audioOptions;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.EndUserId), out string? endUserId))
-                {
-                    result.EndUserId = endUserId;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.IncludeLogProbabilities), out bool includeLogProbabilities))
-                {
-                    result.IncludeLogProbabilities = includeLogProbabilities;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.LogitBiases), out IDictionary<int, int>? logitBiases))
-                {
-                    foreach (KeyValuePair<int, int> kvp in logitBiases!)
-                    {
-                        result.LogitBiases[kvp.Key] = kvp.Value;
-                    }
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.Metadata), out IDictionary<string, string>? metadata))
-                {
-                    foreach (KeyValuePair<string, string> kvp in metadata)
-                    {
-                        result.Metadata[kvp.Key] = kvp.Value;
-                    }
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.OutputPrediction), out ChatOutputPrediction? outputPrediction))
+                if (tool is AIFunction af)
                 {
-                    result.OutputPrediction = outputPrediction;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.ReasoningEffortLevel), out ChatReasoningEffortLevel reasoningEffortLevel))
-                {
-                    result.ReasoningEffortLevel = reasoningEffortLevel;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.ResponseModalities), out ChatResponseModalities responseModalities))
-                {
-                    result.ResponseModalities = responseModalities;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.StoredOutputEnabled), out bool storeOutputEnabled))
-                {
-                    result.StoredOutputEnabled = storeOutputEnabled;
-                }
-
-                if (additionalProperties.TryGetValue(nameof(result.TopLogProbabilityCount), out int topLogProbabilityCountInt))
-                {
-                    result.TopLogProbabilityCount = topLogProbabilityCountInt;
+                    result.Tools.Add(ToOpenAIChatTool(af));
```

Comment:
Does this risk adding more and more copies of the same tool every time the same `options` object is used?

Reply:
Yes, it does.

Reply:
We can't just clear it out first, as someone could reasonably expect that they can add tools here, especially one's that the abstraction can't yet represent. And we can't just check first whether the tool is there, because if the same object is being reused, it'll include tools from the previous go around that were never removed, and thus this could include tools not intended for reuse. Seems like we're going to have to clone, which is unfortunate.